pkg/config/crd/crdwatcher.go

```diff
@@ -0,0 +1,273 @@
+type resourceMap map[string]*resource
+
+type watcher struct {
+	namespace string
+	stoppers  []func()
+
+	mu       sync.Mutex
+	data     map[string]string
+	listener store.Listener
+
+	// kind to resourceMap
+	resources map[string]resourceMap
+}
+
+var _ store.ChangeNotifier = &watcher{}
+
+// newStore creates a new watcher instance with watching resources.
+func newStore(u *url.URL) (store.KeyValueStore, error) {
+	namespace := u.Host
+	if namespace == "" {
+		namespace = v1.NamespaceDefault
+	}
+	configFilePath := u.Path
+	if configFilePath == "" {
+		configFilePath = filepath.Join(os.Getenv("HOME"), ".kube", "config")
+	}
+	config, err := clientcmd.BuildConfigFromFlags("", configFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	w := &watcher{
+		data:      map[string]string{},
+		namespace: namespace,
+		resources: map[string]resourceMap{},
+	}
+	if err = w.init(config); err != nil {
+		return nil, err
+	}
+	return w, nil
+}
+
+func (w *watcher) String() string {
+	return fmt.Sprintf("k8s: %s", w.namespace)
+}
+
+// Get implements a KeyValueStore method.
+func (w *watcher) Get(key string) (string, int, bool) {
+	w.mu.Lock()
+	value, found := w.data[key]
+	w.mu.Unlock()
+	return value, store.IndexNotSupported, found
+}
+
+// Set implements a KeyValueStore method.
+func (w *watcher) Set(key, value string) (int, error) {
+	return store.IndexNotSupported, errors.New("not implemented")
+}
+
+// List implements a KeyValueStore method.
+func (w *watcher) List(key string, recurse bool) ([]string, int, error) {
+	w.mu.Lock()
+	if !strings.HasSuffix(key, "/") {
+		key = key + "/"
+	}
+	keys := []string{}
+	for k := range w.data {
+		if strings.HasPrefix(k, key) {
+			keys = append(keys, k)
+		}
+	}
+	w.mu.Unlock()
+	return keys, store.IndexNotSupported, nil
+}
+
+// Delete implements a KeyValueStore method.
+func (w *watcher) Delete(key string) error {
+	return errors.New("not implemented")
+}
+
+func unstructuredToResource(uns *unstructured.Unstructured) (string, *resource) {
+	md := uns.Object["metadata"].(map[string]interface{})
+	name := md["name"].(string)
+	version := md["resourceVersion"].(string)
+	return name, &resource{version, uns.Object["spec"]}
+}
+
+func (w *watcher) appendAndEncode(data resourceMap, field, key string) {
+	if len(data) == 0 {
+		return
+	}
+	res := make([]interface{}, 0, len(data))
+	for _, d := range data {
+		res = append(res, d.spec)
+	}
+	enc, err := yaml.Marshal(map[string]interface{}{field: res})
+	if err != nil {
+		glog.Errorf("Failed to marshal: %v", err)
+		return
+	}
+	w.data[key] = string(enc)
+}
+
+func (w *watcher) encodeData(data resourceMap, field string) {
+	for name, d := range data {
+		key := "/scopes/global/subjects/" + name + "/" + field
+		enc, err := yaml.Marshal(map[string]interface{}{field: d.spec})
+		if err != nil {
+			glog.Errorf("Failed to marshal: %v", err)
+			return
+		}
+		w.data[key] = string(enc)
+	}
+}
+
+func (w *watcher) buildDataAndNotify() {
+	w.mu.Lock()
```

Comment:
If `resources` is not guarded by `mu` then this should be re-written - you could hold `mu` for a substantially shorter time period (just the updates at the end of `appendAndEncode`, `encodeData`) while still making the update look atomic because you don't notify listeners until construction is finished.